### ISSUE TITLE
Windows Sound Recorder MP3 ID3 tag breaks playback

### ISSUE THREAD

**Issue by toybuilder** [chmorgan/esp-libhelix-mp3#1]
I encountered a problem while playing back MP3 files created through Windows Sound Recorder when playing through the ESP_I2S playMP3() call.  See https://github.com/espressif/arduino-esp32/issues/10469

The call results in this error:
   `[1850322][E][ESP_I2S.cpp:1035] playMP3(): Decode ERROR: -6`
Which corresponds to ERR_MP3_INVALID_FRAMEHEADER

I am pasting in the hexdump of the beginning of the problem MP3 file here.  The full file can be downloaded from https://github.com/toybuilder/esp-arduino-i2s-audio/

I think the problem might be due to the byte pattern 0xff 0xfe at 0x0015 being interpreted as a sync value when the correct sync is (I believe) at 0x1af.

If I am understanding the ID3v2 tag definition (https://mutagen-specs.readthedocs.io/en/latest/id3/id3v2.4.0-structure.html) correctly, the ID3 header defines a length of 0x1a5 bytes (0x0325 at location 0x6-0x9, with MSB of each octet discarded resulting in 0x1a5), putting the start of the MP3 stream at 0x1af (0x1a5 + 10).

```

00000000: 4944 3303 0000 0000 0325 5446 4c54 0000  ID3......%TFLT..
00000010: 000f 0000 01ff fe4d 0050 0047 002f 0033  .......M.P.G./.3
00000020: 0000 0043 4f4d 4d00 0000 8200 0000 656e  ...COMM.......en
00000030: 6769 5475 6e53 4d50 4200 2030 3030 3030  giTunSMPB. 00000
00000040: 3030 3020 3030 3030 3030 3030 2030 3030  000 00000000 000
00000050: 3030 3030 3020 3030 3030 3030 3030 3030  00000 0000000000
00000060: 3032 6133 3166 2030 3030 3030 3030 3020  02a31f 00000000 
00000070: 3030 3030 3030 3030 2030 3030 3030 3030  00000000 0000000
00000080: 3020 3030 3030 3030 3030 2030 3030 3030  0 00000000 00000
00000090: 3030 3020 3030 3030 3030 3030 2030 3030  000 00000000 000
000000a0: 3030 3030 3020 3030 3030 3030 3030 0000  00000 00000000..
000000b0: 0000 0000 0000 0000 0000 0000 0000 0000  ................
000000c0: 0000 0000 0000 0000 0000 0000 0000 0000  ................
000000d0: 0000 0000 0000 0000 0000 0000 0000 0000  ................
000000e0: 0000 0000 0000 0000 0000 0000 0000 0000  ................
000000f0: 0000 0000 0000 0000 0000 0000 0000 0000  ................
00000100: 0000 0000 0000 0000 0000 0000 0000 0000  ................
00000110: 0000 0000 0000 0000 0000 0000 0000 0000  ................
00000120: 0000 0000 0000 0000 0000 0000 0000 0000  ................
00000130: 0000 0000 0000 0000 0000 0000 0000 0000  ................
00000140: 0000 0000 0000 0000 0000 0000 0000 0000  ................
00000150: 0000 0000 0000 0000 0000 0000 0000 0000  ................
00000160: 0000 0000 0000 0000 0000 0000 0000 0000  ................
00000170: 0000 0000 0000 0000 0000 0000 0000 0000  ................
00000180: 0000 0000 0000 0000 0000 0000 0000 0000  ................
00000190: 0000 0000 0000 0000 0000 0000 0000 0000  ................
000001a0: 0000 0000 0000 0000 0000 0000 0000 00ff  ................
000001b0: fbb4 4000 0002 5708 20f9 8919 205a 20f4  ..@...W. ... Z .
000001c0: 1d31 8214 0c2d 6d13 8798 4da9 420e e2b0  .1...-m...M.B...
000001d0: f321 2488 7666 5432 004a 72c9 80f0 1906  .!$.vfT2.Jr.....
000001e0: 4188 cc58 2712 09cd 9e68 61c6 704c 4c64  A..X'....ha.pLLd
000001f0: cb90 68d1 a0b0 84d1 a346 ddad a64c 9930  ..h......F...L.0
00000200: 68d1 a0b3 3346 cc99 32ef f86c c993 2cf3  h...3F..2..l..,.

```

I am not sure if the responsibility of parsing out and skipping the ID3 tag should be in your library, or if this behavior needs to be documented so that the users can take the appropriate steps.  If the latter, then I should and could supply my own code to look for the ID3 tag.   

Thank you.

**Comment by chmorgan** [comment]
So the library is partially used to aid in finding the next frame. If you look at this code you can see how it works,

https://github.com/chmorgan/esp-audio-player/blob/main/audio_mp3.cpp

effectively even if there is a decode error there is an attempt to find the next sync and continue processing data.

Thoughts?

**Comment by toybuilder** [comment]
If I'm understanding your code correctly, your is_mp3() will seek the file back to the beginning, so later call to decode_mp3() would result in the same problem, where the ID3 header is not skipped over, resulting in a wrong sync pattern detection. 

So the wrong offset is retrieved at https://github.com/chmorgan/esp-audio-player/blob/916bc82aaa7a342fb1f4b908bc66df89055ce70e/audio_mp3.cpp#L89 

And that wrong offset is used for the playback attempt https://github.com/chmorgan/esp-audio-player/blob/916bc82aaa7a342fb1f4b908bc66df89055ce70e/audio_mp3.cpp#L96C1-L100C1

which results in the decoding the MP3 data from the middle of an ID3 tag, which results in ERR_MP3_INVALID_FRAMEHEADER

and there is no looping to seek the next sync to continue playback.

The main issue here is that it's possible that an ID3 tag contains a byte sequence that looks like a sync, and that is what's breaking the playback.

**Comment by toybuilder** [comment]
Specifically, these bytes in the problematic ID3 header:

00000010: 000f 0000 01**ff** **fe**4d 0050 0047 002f 0033  .......M.P.G./.3

**Comment by chmorgan** [comment]
If it wasn’t clear, if you find the next sync marker and decode again, it will eventually find the correct one and you’ll be off and running. The code above (decode_mp3()) is called repeatedly, feeling new data in to decode, that’s why you don’t see the loop itself there. It’s the call to find sync word in the decode_mp3().

It’s not the most ideal. Perhaps it makes sense to see if it’s a header and skip those bytes since we can find the length, but if you wanted what I think is an easy fix you can have your code do something like:

```
while(data)
{
    MP3FindSyncWord
    decode at position
}
```

That’s what I do with esp-audio-player.

**Comment by chmorgan** [comment]
And the is mp3 stuff is not intended to be called each time you decode_mp3(), so it only ever seeks forward.

**Comment by toybuilder** [comment]
Oh, ok, I now see how your implementation will eventually find a sync again.  I didn't fully appreciate how audio_player.c will loop calling decode_mp3() with the same decoder instance to continue where it last left off until eof is reached.

In the ESP_I2S playMP3() implementation, there is no attempt to do so.  

I think there is a disconnect between implied/expected behavior and actual implementation there, and that's why I ended up down this rabbit hole.  Hopefully, this discussion is enough for future reference for the next person that might have the same issue.

**Comment by chmorgan** [comment]
@toybuilder we could update the player docs to be explicit about what the user should do next, I don't think we have even a pseudo code example of how to use the decode function.

**Comment by toybuilder** [comment]
I think that bit of warning would go a long way to reduce confusion for others in the future...

**Comment by chmorgan** [comment]
@toybuilder would a pointer in the readme to the esp-audio-player be sufficient to show how to use it? Or some pseudocode in the readme.md? What do you think would have been most helpful?

**Comment by toybuilder** [comment]
Perhaps something like:

NOTE: MP3Decode() only handles the decoding of the MP3 stream contained within the supplied buffer.  If the supplied data contains an ID3 tag header, the decoding may return prematurely if the ID3 data contain bytes that resembles an MP3 synchronization data pattern.    Removing the ID3 data from the source data, or calling MP3Decode() again to resume processing may be necessary.  See esp-audio-player(link) for an example of how to do this.

**Comment by chmorgan** [comment]
@toybuilder see #2

**Comment by chmorgan** [comment]
Merged #2.
